app/models/composition_player.rb

```diff
@@ -0,0 +1,31 @@
+class CompositionPlayer < ApplicationRecord
+  belongs_to :composition
+  belongs_to :player
+
+  before_validation :set_position
+
+  validates :player, :composition, :position, presence: true
+  validates :position, numericality: { only_integer: true }
+  validate :composition_does_not_have_max_players
+
+  default_scope { order(:position) }
+
+  private
+
+  def set_position
+    return unless composition
+
+    self.position ||= composition.composition_players.count
```

Comment:
I'm not sure we want to memoize here. This seems to have adverse affects on the position calculated for new players added to a composition. When memoized, the `CompositionPlayer` records created for me always have a position of `0`. Was there a specific goal you had in mind with memoizing here?